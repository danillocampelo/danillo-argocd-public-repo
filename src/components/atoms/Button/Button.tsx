import React, { ComponentClass, FunctionComponent } from 'react'
import { Themes } from '@utils/constants/theme'
import PrimaryButton from './components/PrimaryButton'
import SecondaryButton from './components/SecondaryButton'
import TertiaryButton from './components/TertiaryButton'
import QuaternaryButton from './components/QuaternaryButton'

export type ButtonTypes = 'primary' | 'secondary' | 'tertiary' | 'quaternary'

export type Size = 'extra-small' | 'small' | 'regular'

export type Icon =
  | string
  | FunctionComponent<{ width: number; height: number }>
  | ComponentClass<{ width: number; height: number }, any>

export interface TypedButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  theme: Themes
  children?: React.ReactNode
}

export interface Props extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  buttonType: ButtonTypes
  children?: React.ReactNode
  Icon?: Icon
  theme?: Themes
  typography?: string
  className?: string
  htmlType?: 'submit' | 'button' | 'reset'
  reverseIcon?: boolean
  hideTextOnMobile?: boolean
}

const Button = ({
  buttonType,
  children,
  Icon,
  theme = Themes.light,
  className = 'button-regular',
  typography = 'text-button',
  reverseIcon = false,
  hideTextOnMobile = false,
  ...ButtonProps
}: Props) => {
  const defaultClass = `flex items-center justify-center whitespace-nowrap`

  const iconPositionClass = Icon
    ? reverseIcon
      ? hideTextOnMobile
        ? 'xl:has-right-icon'
        : 'has-right-icon'
      : hideTextOnMobile
      ? 'has-left-icon'
      : 'xl:has-left-icon'
    : ''

  const TypedButton = ({ theme, className, children }: TypedButtonProps) => {
    switch (buttonType) {
      case 'primary':
        return (
          <PrimaryButton theme={theme} {...ButtonProps} className={className}>
            {children}
          </PrimaryButton>
        )
      case 'secondary':
        return (
          <SecondaryButton theme={theme} {...ButtonProps} className={className}>
            {children}
          </SecondaryButton>
        )
      case 'tertiary':
        return (
          <TertiaryButton theme={theme} {...ButtonProps} className={className}>
            {children}
          </TertiaryButton>
        )
      case 'quaternary':
        return (
          <QuaternaryButton
            theme={theme}
            {...ButtonProps}
            className={className}
          >
            {children}
          </QuaternaryButton>
        )
    }
  }

  return (
    <TypedButton
      theme={theme}
      className={`${defaultClass} ${typography} ${className} ${
        reverseIcon ? 'flex-row-reverse' : ''
      } ${iconPositionClass}`}
      {...ButtonProps}
    >
      {Icon && <Icon width={24} height={24} />}
      {children && (
        <span className={`${hideTextOnMobile ? 'hidden lg:block' : ''}`}>
          {children}
        </span>
      )}
    </TypedButton>
  )
}

export default Button
