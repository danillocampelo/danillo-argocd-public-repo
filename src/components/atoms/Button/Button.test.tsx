import React from 'react'
import { render, screen } from '@testing-library/react'
import Button from './Button'

describe('Button', () => {
  it('should render with default styles', () => {
    render(<Button buttonType="primary">Default</Button>)

    expect(screen.getByText('Default')).toMatchSnapshot()
  })

  it('should render with disabled property', () => {
    render(
      <Button buttonType="primary" disabled>
        Disabled Button
      </Button>,
    )

    expect(screen.getByText('Disabled Button')).toHaveProperty('disabled', true)
  })
})
