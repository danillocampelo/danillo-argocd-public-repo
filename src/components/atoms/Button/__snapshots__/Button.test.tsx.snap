// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render with default styles 1`] = `
<button
  class="flex uppercase py-4 px-6 mr-0 font-medium leading-4 font-tertiary justify-around items-center w-[164px] border border-secondary bg-secondary text-secondary-contrast disabled:bg-gray-800 hover:text-secondary hover:bg-secondary-contrast hover:border-secondary undefined undefined"
>
  Default
</button>
`;
